atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCells.java

```diff
@@ -156,54 +157,65 @@ SweepBatch getBatchForPartition(ShardAndStrategy shardStrategy, long partitionFi
-                .flatMap(Collection::stream)
-                .forEach(write -> writesToSweepFor
-                        .putIfAbsent(write.writeRef().cellReference(), write));
-        return writesToSweepFor.values();
-    }
-
-    private long getLastSweptTs(Multimap<Long, WriteInfo> writesByStartTs, RowColumnRangeIterator resultIterator,
-            long partitionFine, long maxTsExclusive) {
-        if (exhaustedAllColumns(resultIterator)) {
-            return lastGuaranteedSwept(partitionFine, maxTsExclusive);
-        }
-        return writesByStartTs.keySet().stream().max(Comparator.naturalOrder()).orElse(-1L);
-    }
-
     private RowColumnRangeIterator getRowColumnRange(SweepableCellsTable.SweepableCellsRow row, long partitionFine,
             long minTsExclusive, long maxTsExclusive) {
         return getRowsColumnRange(ImmutableList.of(row.persistToBytes()),
                 columnsBetween(minTsExclusive + 1, maxTsExclusive, partitionFine), SweepQueueUtils.MAX_CELLS_DEDICATED);
     }
 
-    private List<Long> getCommittedTimestampsDescendingAndCleanupAborted(
-            ShardAndStrategy shardStrategy, Multimap<Long, WriteInfo> startTsWrites) {
-        Map<Long, Long> startToCommitTs = commitTsCache.loadBatch(startTsWrites.keySet());
+    private TimestampsToSweep getTimestampsToSweepDescendingAndCleanupAborted(ShardAndStrategy shardStrategy,
+            long minTsExclusive, long sweepTs, Multimap<Long, WriteInfo> writesByStartTs) {
+        Map<Long, Long> startToCommitTs = commitTsCache.loadBatch(writesByStartTs.keySet());
         Map<TableReference, Multimap<Cell, Long>> cellsToDelete = new HashMap<>();
         List<Long> committedTimestamps = new ArrayList<>();
-
-        for (Map.Entry<Long, Long> entry : startToCommitTs.entrySet()) {
-            if (entry.getValue() == TransactionConstants.FAILED_COMMIT_TS) {
-                startTsWrites.get(entry.getKey())
+        long lastSweptTs = minTsExclusive;
+        boolean processedAll = true;
+
+        List<Long> sortedStartTimestamps = startToCommitTs.keySet().stream().sorted().collect(Collectors.toList());
+        for (long startTs : sortedStartTimestamps) {
+            long commitTs = startToCommitTs.get(startTs);
+            if (commitTs == TransactionConstants.FAILED_COMMIT_TS) {
+                lastSweptTs = startTs;
+                writesByStartTs.get(startTs)
                         .forEach(write -> cellsToDelete
                                 .computeIfAbsent(write.tableRef(), ignore -> HashMultimap.create())
                                 .put(write.cell(), write.timestamp()));
+            } else if (commitTs < sweepTs) {
+                lastSweptTs = startTs;
+                committedTimestamps.add(startTs);
             } else {
-                committedTimestamps.add(entry.getKey());
+                processedAll = false;
+                break;
             }
         }
+
         cellsToDelete.forEach((tableRef, multimap) -> {
```

Comment:
For future note, I'm curious how often we need to delete an aborted txn's writes directly vs. the batch containing a valid write already that would cover it with a ranged tombstone. If batch sizes are sufficiently high, we might be able to avoid most of the kvs.delete calls here if each sweep iteration needs to be sped up.

Reply:
Good point